Python/ceval_gil.c

```diff
@@ -57,113 +57,60 @@
 #define _Py_atomic_load_relaxed_int32(ATOMIC_VAL) _Py_atomic_load_relaxed(ATOMIC_VAL)
 #endif
 
-/* This can set eval_breaker to 0 even though gil_drop_request became
-   1.  We believe this is all right because the eval loop will release
-   the GIL eventually anyway. */
+
 static inline void
-COMPUTE_EVAL_BREAKER(PyInterpreterState *interp,
-                     struct _ceval_runtime_state *ceval,
-                     struct _ceval_state *ceval2)
+update_eval_breaker_from_thread(PyInterpreterState *interp, PyThreadState *tstate)
 {
-    _Py_atomic_store_relaxed(&ceval2->eval_breaker,
-        _Py_atomic_load_relaxed_int32(&ceval2->gil_drop_request)
-        | (_Py_atomic_load_relaxed_int32(&ceval->signals_pending)
-           && _Py_ThreadCanHandleSignals(interp))
-        | (_Py_atomic_load_relaxed_int32(&ceval2->pending.calls_to_do))
-        | (_Py_IsMainThread() && _Py_IsMainInterpreter(interp)
-           &&_Py_atomic_load_relaxed_int32(&ceval->pending_mainthread.calls_to_do))
-        | ceval2->pending.async_exc
-        | _Py_atomic_load_relaxed_int32(&ceval2->gc_scheduled));
+    if (tstate == NULL) {
+        return;
+    }
+    int32_t calls_to_do;
+    if (_Py_ThreadCanHandleSignals(interp)) {
+        calls_to_do = _Py_atomic_load_int32_relaxed(
+            &_PyRuntime.ceval.pending_mainthread.calls_to_do);
+        if (_Py_atomic_load(&_PyRuntime.signals.is_tripped)) {
+            _Py_set_eval_breaker_bit(interp, _PY_SIGNALS_PENDING_BIT, 1);
+        }
+    }
+    else {
+        calls_to_do = _Py_atomic_load_int32_relaxed(
+            &interp->ceval.pending.calls_to_do);
```

Comment:
Yes it is.

In earlier version, I set the bit to zero or one, depending on the calls to do. In which case it could be set from 1 to 0 when switching from the main thread.
Now, we only set to 1 as it is faster, so this isn't needed.